### ISSUE TITLE
UIowa Sections: Use blocks instead of paragraphs?

### ISSUE THREAD

**Issue by pyrello** [uiowa/collegiate#1]
Note: This issue will eventually live in the repo for the uiowa_sections module, but is going to live here until that is created.

This is a place to keep track of any discussion/notes/resources around deciding whether to go with a paragraphs-type approach similar to what the sitenow profile uses, or to possibly use custom blocks + block fields + layout builder to accomplish something similar.

Here's is what I understand so far:
* Using Entity References (and I'm assuming possibly Entity Reference Revisions) fields, you can reference plugin blocks (config) or custom blocks (content) but not both with the same field.
* There is a field that supersedes the blockreference module called block_field, which I am going to test out.
* Custom blocks have some sort of revisioning in place or in progress, but no UI exists for it yet. The block revisions are supposed to be an 8.7 feature and the UI to support them is possibly an 8.8 feature.
* The layout builder module has some way of working with custom blocks that needs some additional investigation. I'm not sure if it include revisions or not.

Based on the above, it seems that if the block_field module doesn't provide a solution which provides revisions based on parent entity updates, then this may be a non-starter.

**Comment by bspeare** [comment]
What type of UI are they planning with the custom block revisions in 8.8?  I just tested out 8.7.0-alpha2 and was able to revert each update that I made to custom blocks through layout builder.

**Comment by pyrello** [comment]
This is the thread I was looking at. I didn't get too in depth with it though.

https://www.drupal.org/project/drupal/issues/1984588

Maybe that is just a feature request for revisions UI for blocks not in layout builder.

I'll have to play around with layout builder a little bit more.

**Comment by pyrello** [comment]
It seems that, unless I am missing something, one of the inherent downsides to using blocks + layout builder is that, while it works pretty well for existing content, the workflow is still a little clunky for creating new content. Assuming that we could build out a way to do something similar to the sitenow paragraphs implementation - where we have custom block types to allow people to add different types of widgets - we would essentially have to have people create a blank node, then go to edit the layout. 
<img width="1002" alt="Screen Shot 2019-03-27 at 3 50 06 PM" src="https://user-images.githubusercontent.com/709009/55111343-2b23b900-50a8-11e9-857b-e403bd4a22fd.png">
<img width="1309" alt="Screen Shot 2019-03-27 at 3 50 55 PM" src="https://user-images.githubusercontent.com/709009/55111346-2e1ea980-50a8-11e9-80c1-8bef1db06579.png">

Not ideal.

**Comment by pyrello** [comment]
@bspeare I'm not sure how to get to the block revision you were able to do. Can you provide steps?

**Comment by bspeare** [comment]
It's good to see that the custom blocks are heading towards a revisions UI, but being able to revert custom block changes through layout builder is a great feature.

**Comment by bspeare** [comment]
@pyrello I used https://simplytest.me/project/drupal with 8.7.0-alpha2.

**Comment by pyrello** [comment]
Oh... gotcha. I was trying to do it against 8.6.whatever.

**Comment by bspeare** [comment]
I agree that the workflow is clunky but at least it provides a visual of the layout instead of digging down into nested paragraphs and selecting layout classes.   People are a little used to that workflow with sitenow panels so it wouldn't be that much of a stretch.

**Comment by bspeare** [comment]
It's also a workflow used in most Wordpress layout builders (Elementor) so it is a pretty common pattern.

**Comment by pyrello** [comment]
I set up a test installation of Drupal 8.7.0-alpha2 to checkout the revision stuff. It works pretty well, at least from the frontend.

To test the viability of the blocks + layout approach to complex section building, I attempted to create a simple example scenario: Being able to add an accordion and its child items. To do this, I created a new custom block type called Accordion. I removed the body field and added a reference field which allows the user to select an unlimited number of basic blocks. Since basic blocks consist of a title and a body field, this works well for an item in an accordion since those would be the fields we would need.

The first major issue is there is no built-in form display widget for a nested block that allows you to actually create the block in that context. All the built-in entity reference widgets assume that the block already exists (e.g. select list, autocomplete, etc.). It might be possible to use something like IEF (https://www.drupal.org/project/inline_entity_form) to show a creation form.

Another potential issue that comes out of this nested block situation is that the mechanism that manages whether blocks are local to the context of an entity (via layout builder) or global likely won't see nested blocks as belonging to the parent entity. So my guess is that any block created that way will show up in the global block list. Testing this theory will require using something like IEF to be able to create a nested block in the first place.

A possible short-term workaround is to use a block + paragraphs + layout builder approach: All top level "widgets" that can be added to the layout are blocks, but any nested references reference paragraph types instead. So for my example above, I would have a custom block type called Accordion, but the field_accordion_items field would reference (via ERR) a paragraph type called item. Since paragraphs are revision-able already, I would assume that the revision settings would just work, although this is definitely theoretical and would also need to be tested. There is a lot of black box magic going on with the layout field, which makes parsing out what is happening more difficult.

Finally, since the revision-able inline blocks seems to be a 8.7.x feature, that means we are not going to have a stable release until May 1. We could potentially develop against 8.7 until then, so maybe that wouldn't be an issue.

**Comment by bspeare** [comment]
I just tried your short-term workaround with block + paragraphs + layout builder approach and it seems to work well.   Since we have a smaller amount of sites it would be nice if we could try to move forward with a layout builder solution before we revert to the uiowa_paragraphs.  What do you think?  Is there anything in particular you want to start testing?

**Comment by pyrello** [comment]
Did you have a chance to try out revisions? Did that work as you would expect?

**Comment by bspeare** [comment]
Not exactly.  Perhaps this is the issue: https://youtu.be/jKh8HuECm3g?t=650.   Paragraphs referenced within a block or content type using ERR have to be updated separately in order to receive the updated content from the source.  Have you noticed that aspect?

**Comment by bspeare** [comment]
One way around that is to have each paragraph as a separate content field in the content type so that you can place the paragraph as field through the layout builder settings tray instead of via a custom block type that uses ERR.  

<img width="1122" alt="Screen Shot 2019-03-28 at 2 17 29 PM" src="https://user-images.githubusercontent.com/1036433/55186342-4eac3980-5164-11e9-9bf0-5ed5230953f4.png">

**Comment by pyrello** [comment]
I could not even get mine to save, I encountered the error described in this issue: https://www.drupal.org/project/paragraphs/issues/2901390

**Comment by bspeare** [comment]
Is that where you are creating the paragraph through the custom block?  Or are you just linking to the paragraph through the node?

**Comment by pyrello** [comment]
I created a paragraph type called Item with title and body fields. I created a custom block type called Accordion which uses an ERR field to reference the Item paragraph type. I created a block via the "Create custom block" on the layout builder. When I went to save the layout, I got the error.

**Comment by pyrello** [comment]
I think the issue is more similar than what I was originally thinking. It seems like it might be happening in that case because IEF is not using the normal Entity Form, which is where the langcode is set.

**Comment by bspeare** [comment]
That patch (https://www.drupal.org/files/issues/2019-03-22/2901390-n24.patch) fixes the issue.  Did you try it?

**Comment by bspeare** [comment]
I tested Paragraph revisions with Layout Builder in 8.7.0-beta1 using the above patch and I could not get the revisions to work.  The page reverts for regular content updates but the paragraph content disappears and the message "You are not allowed to edit or remove this Paragraph" appears where the paragraph fields were.

**Comment by pyrello** [comment]
So, to clarify, the patch fixes the issue with the language not being set, but inline custom blocks + paragraphs does not work with revisions as expected?

Unless I am missing something, if we want to make the blocks route work, this would seem to leave us with the approach you outlined above, @bspeare, where we would add every "widget" field type to each content type and then the fields can be placed via layout. I'm going to give this a go to see what the user experience is like and I'll report back with my thoughts.
